components_library/boards/lupv_board.py

```diff
@@ -0,0 +1,389 @@
+
+        # LUPIO BLK
+        self.lupio_blk = LupioBLK(pio_addr=0x20000000, pic = self.pic)
+
+        # Note: This only works with single threaded cores.
+        self.pic.n_contexts = self.processor.get_num_cores() * 2
+        # Interrupt IDs for LupIO devices
+        pic_srcs = [
+            self.lupio_tty.int_id,
+            self.lupio_blk.int_id,
+            self.lupio_rng.int_id
+        ]
+        self.pic.n_src = max(pic_srcs) + 1
+
+        self.clint.num_threads = self.processor.get_num_cores()
+        # Add the RTC
+        # TODO: Why 100MHz? Does something else need to change when this does?
+        self.rtc = RiscvRTC(frequency=Frequency("100MHz"))
+        self.clint.int_pin = self.rtc.int_pin
+
+        # Incoherent I/O bus
+        self.iobus = IOXBar()
+
+        # Note: This overrides the platform's code because the platform isn't
+        # general enough.
+        self._on_chip_devices = [self.clint, self.pic]
+        self._off_chip_devices = [
+            self.lupio_blk,
+            self.lupio_rtc,
+            self.lupio_rng,
+            self.lupio_tty
+        ]
+
+    def _setup_io_devices(self) -> None:
+        """Connect the I/O devices to the I/O bus"""
+        for device in self._off_chip_devices:
+            device.pio = self.iobus.mem_side_ports
+        self.lupio_blk.dma = self.iobus.cpu_side_ports
+
+        for device in self._on_chip_devices:
+            device.pio = self.get_cache_hierarchy().get_mem_side_port()
+        self.bridge = Bridge(delay="10ns")
+        self.bridge.mem_side_port = self.iobus.cpu_side_ports
+        self.bridge.cpu_side_port = (
+            self.get_cache_hierarchy().get_mem_side_port()
+        )
+        self.bridge.ranges = [
+            AddrRange(dev.pio_addr, size=dev.pio_size)
+            for dev in self._off_chip_devices
+        ]
+
+    def _setup_pma(self) -> None:
+        """Set the PMA devices on each core"""
+        uncacheable_range = [
+            AddrRange(dev.pio_addr, size=dev.pio_size)
+            for dev in self._on_chip_devices + self._off_chip_devices
+        ]
+        # TODO: Not sure if this should be done per-core like in the example
+        for cpu in self.get_processor().get_cores():
+            cpu.get_mmu().pma_checker = PMAChecker(
+                uncacheable=uncacheable_range
+            )
+
+    @overrides(AbstractBoard)
+    def has_io_bus(self) -> bool:
+        return True
+    @overrides(AbstractBoard)
+    def get_io_bus(self) -> IOXBar:
+        return self.iobus
+    def has_coherent_io(self) -> bool:
+        return True
+    def get_mem_side_coherent_io_port(self) -> Port:
+        return self.iobus.mem_side_ports
+    @overrides(AbstractBoard)
+    def setup_memory_ranges(self):
+        memory = self.get_memory()
+        mem_size = memory.get_size()
+        self.mem_ranges = [AddrRange(start=0x80000000, size=mem_size)]
+        memory.set_memory_range(self.mem_ranges)
+    def set_workload(
```

Comment:
I think Python Black needs run here. There should be a space between these functions. Not a big deal but it looks a little ugly.